File: src/Prompts/Functions/DebtAnalysis.php
<?php

declare(strict_types=1);

namespace App\Prompts\Functions;

use App\PromptLoader;
use App\Prompts\AnalysisFocus;
use App\Prompts\Core\OutputFormats;
use App\Prompts\Enums\Focus;
use App\Prompts\Enums\Language;
use App\Prompts\Enums\Perspective;

/**
 * 技術的負債分析プロンプト
 */
final class DebtAnalysis
{
    /**
     * 技術的負債分析プロンプトを生成
     *
     * @param string $code 分析対象のコード
     * @param string|null $perspective 設計観点（ddd, laravel, clean）
     * @param string|null $language プログラミング言語（php, typescript）
     * @param list<string> $focuses 分析観点のリスト（グループまたは個別観点）
     */
    public static function generate(
        string $code,
        ?string $perspective = null,
        ?string $language = null,
        array $focuses = []
    ): string {
        $loader = PromptLoader::getInstance();

        // focusesが空の場合はデフォルト
        if ($focuses === []) {
            $focuses = AnalysisFocus::defaults();
        }

        // 観点を解決
        $resolved = AnalysisFocus::resolve($focuses);

        // coreプロンプトを構築
        $corePrompt = self::buildPromptFromFocuses($resolved['core']);

        // antipatternsプロンプトを構築
        $antipatternsPrompt = self::buildPromptFromFocuses($resolved['antipatterns']);

        $outputFormat = OutputFormats::all();

        // 設計観点（Perspective）
        $perspectivePrompt = '';
        if ($perspective !== null) {
            $perspectiveEnum = Perspective::fromAlias($perspective);
            if ($perspectiveEnum !== null) {
                $perspectivePrompt = $loader->getContent($perspectiveEnum->promptPath());
            }
        }

        // 言語（Language）
        $languagePrompt = '';
        if ($language !== null) {
            $languageEnum = Language::fromAlias($language);
            if ($languageEnum !== null) {
                $languagePrompt = $loader->getContent($languageEnum->promptPath());
            }
        }

        return $loader->renderTemplate('functions/debt-analysis/base', [
            'corePrompt' => $corePrompt,
            'antipatternsPrompt' => $antipatternsPrompt,
            'perspectivePrompt' => $perspectivePrompt,
            'languagePrompt' => $languagePrompt,
            'code' => $code,
            'outputFormat' => $outputFormat,
        ]);
    }

    /**
     * Focusリストからプロンプトを構築
     *
     * @param list<Focus> $focuses
     */
    private static function buildPromptFromFocuses(array $focuses): string
    {
        if ($focuses === []) {
            return '';
        }

        $loader = PromptLoader::getInstance();
        $contents = [];

        foreach ($focuses as $focus) {
            $path = $focus->promptPath();
            if ($loader->exists($path)) {
                $contents[] = $loader->getContent($path);
            }
        }

        return implode("\n\n---\n\n", $contents);
    }
}

File: functions/debt_analysis.go
package functions

import (
	"seiren/promptloader"
	"seiren/domain"
	"strings"
)

func GenerateDebtAnalysis(code, perspective, language string, focuses []string) string {
	loader := promptloader.GetInstance()

	if len(focuses) == 0 {
		focuses = domain.DefaultFocuses()
	}

	resolved := domain.ResolveFocuses(focuses)

	corePrompt := buildPromptFromFocuses(resolved.Core, loader)
	antipatternsPrompt := buildPromptFromFocuses(resolved.Antipatterns, loader)
	outputFormat := domain.OutputFormatsAll(loader)

	perspectivePrompt := ""
	if perspective != "" {
		if p, ok := domain.PerspectiveFromAlias(perspective); ok {
			perspectivePrompt = loader.GetContent(p.PromptPath())
		}
	}

	languagePrompt := ""
	if language != "" {
		if l, ok := domain.LanguageFromAlias(language); ok {
			languagePrompt = loader.GetContent(l.PromptPath())
		}
	}

	targetCode := buildDebtAnalysisTargetCode(code, loader)

	return loader.RenderTemplate("functions/debt-analysis/base", map[string]string{
		"corePrompt":         corePrompt,
		"antipatternsPrompt": antipatternsPrompt,
		"perspectivePrompt":  perspectivePrompt,
		"languagePrompt":     languagePrompt,
		"outputFormat":       outputFormat,
		"targetCode":         targetCode,
	})
}

func buildDebtAnalysisTargetCode(code string, loader *promptloader.Loader) string {
	if code == "" {
		return loader.GetContent("functions/debt-analysis/target-code-fallback")
	}
	return loader.RenderTemplate("functions/debt-analysis/target-code", map[string]string{
		"code": code,
	})
}

func buildPromptFromFocuses(focuses []domain.Focus, loader *promptloader.Loader) string {
	if len(focuses) == 0 {
		return ""
	}

	contents := make([]string, 0, len(focuses))
	for _, f := range focuses {
		path := f.PromptPath()
		if loader.Exists(path) {
			contents = append(contents, loader.GetContent(path))
		}
	}
	return strings.Join(contents, "\n\n---\n\n")
}
